binzx/otomi

```diff
@@ -0,0 +1,210 @@
+  last_file_change=$(date -u -r $script_full_path "+%Y-%m-%dT%H:%M:%SZ")
+  last_commit_date=$(echo $last_commit_date | tr -d '"')
+  last_commit_date_sec=$(date --date="$last_commit_date" +%s)
+  last_file_change_sec=$(date --date="$last_file_change" +%s)
+
+  if [ "$last_commit_date_sec" -ne "0" ] && [ "$last_commit_date_sec" -gt "$last_file_change_sec" ]; then
+    read -r -p "Newer version is available, do you want to update (yes/No)? " answer
+    case ${answer:0:1} in
+    y | Y)
+      run_self_update
+      ;;
+    *)
+      return 0
+      ;;
+    esac
+  fi
+}
+[ -z $CI ] && check_update
+
+tmp_env=$(mktemp)
+env | grep -v -e ^PATH= -e ^HOME= -e ^USER= -e ^TMPDIR= -e ^LC_ALL= >$tmp_env
+
+cat >>$tmp_env <<EOF
+SHELL=${path_to_exec_shell}
+OTOMI_TAG=${OTOMI_TAG}
+OTOMI_CALLER_COMMAND=${BASH_SOURCE[0]##*/}
+OTOMI_IN_DOCKER=true
+EOF
+
+helm_config="$HOME/.config/helm"
+uname -a | grep -i darwin >/dev/null && helm_config="$HOME/Library/Preferences/helm"
+
+stack_dir='/home/app/stack'
+executable="node --experimental-specifier-resolution=node ${stack_dir}/dist/otomi.js --"
+
+function mount_with_named_volume() {
+  source=$1
+  dest=$2
+
+  if [[ $(uname -s) == Darwin*  ]]; then
+    echo "$source:$dest"
+    return 0
+  fi
+
+  [ -z "$NAMED_VOLUME" ] && export NAMED_VOLUME='otomi-volume'
+  helper_container_name="otomi-volume-helper"
+
+  [ -z "$source" ] && echo "Error in function 'mount_with_named_volume': source path is empty" 1>&2
+
+  { # Create volume and copy over contents
+    docker container inspect $helper_container_name && docker rm $helper_container_name
+    docker container create --name $helper_container_name -v "$NAMED_VOLUME:$dest" hello-world &&
+      cd "$source" && docker cp . "$helper_container_name:$dest" &&
+      docker rm $helper_container_name
+  } >/dev/null 2>&1
+
+  echo "$NAMED_VOLUME:$dest"
+}
+
+stack_volume=""
+tmp_volume=""
+if [[ $base_dir == *"otomi-core"* ]] || [[ $(pwd) == *"otomi-core"* ]]; then
+  if [[ $base_dir == *"otomi-core"* ]]; then
+    stack_dir=$(cd "$base_dir/.." && pwd)
+  elif [[ $(pwd) == *"otomi-core"* ]]; then
+    stack_dir=$(pwd | awk -F'otomi-core' '{print $1"otomi-core"}')
+  fi
+
+  stack_volume="-v $(mount_with_named_volume "$stack_dir" "$stack_dir")"
+  tmp_volume="-v $(mount_with_named_volume /tmp /tmp)"
+
+  if [ -f "${stack_dir}/dist/otomi.js" ]; then # Use development distribution if exists
+    executable="node --experimental-specifier-resolution=node ${stack_dir}/dist/otomi.js --"
+  fi
+  echo "OTOMI_DEV=true" >>$tmp_env
+fi
+
+cmd="${executable} $*"
+if [ "$1" = "bash" ] && [ "$#" = "1" ]; then # If command is "otomi bash"
+  cmd="bash"
+fi
+
+pull_output=$(docker pull $otomi_tools_image 2>&1 >/dev/null)
+status=$?
+if [ "$status" -ne 0 ]; then
+  echo "$pull_output"
+  echo "Could not find an image for the supplied OTOMI_TAG: '${OTOMI_TAG}'"
+  exit $status
+fi
```

Comment:
```suggestion
docker image inspect $otomi_tools_image >/dev/null 2>&1 || {
  pull_output=$(docker pull $otomi_tools_image 2>&1 >/dev/null)
  status=$?
  if [ "$status" -ne 0 ]; then
    echo "$pull_output"
    echo "Could not find an image for the supplied OTOMI_TAG: '${OTOMI_TAG}'"
    exit $status
  fi
} >/dev/null 2>&1
```

Reply:
Why? You want to pull the latest image, even if it exists?

Reply:
I don't want to pull any image if I already have $otomi_tools_image

Reply:
So you get stuck in an old version, doesn't seem desirable

Reply:
What do you mean "old version"? In this scenario I just built the image. So a developer can never choose what image to use?

Reply:
They sure can, use the `OTOMI_TAG` env variable. There you can define a custom image tag.
So you build `otomi/core:sebastiaanzx` for example, then you can run it as:
`OTOMI_TAG=sebastiaanzx binzx/otomi --help`

Reply:
Yes, that is the expected behaviour, but right now it will just pull a new image every time. So even if the image tag is present, it will try to pull it upstream. I think we want the same thing but I don't think this code works for that purpose. 

```bash
otomix bash
Error response from daemon: manifest for otomi/core:kind not found: manifest unknown: manifest unknown
Could not find an image for the supplied OTOMI_TAG: 'kind'
```

Reply:
I'll think of a workaround

Reply:
https://github.com/redkubes/otomi-core/blob/otomi-zx/binzx/otomi#L183

You can set `NOPULL` environment variable, and it will skip over pulling the latest version. I hope that helps?